### PR TITLE
WIP: Use functional options pattern for NewProvisionController

### PR DESCRIPTION
**Pull request by wongma7**
fixes https://github.com/kubernetes-incubator/external-storage/issues/77

WIP, started writing this before the deletion threshold stuff was merged

### CODE REVIEW

**Comment by wongma7** [comment]
@xingzhou please review if you'd like. The first commit is the actual changes to NewProvisionController that will affect you & other users. Second commit is just converting the provisioners in this repo to it.

Now, invocations will look something like this
```go
ctrl := controller.NewProvisionController(
  client,
  provisionerName,
  provisioner,
  serverGitVersion,
  controller.ResyncPeriod(60 * time.Second),
  controller.ExponentialBackOffOnError(false),
)
```
And I have set defaults for many parameters, since before it was mandatory to set them and now theya re optional.

File: lib/controller/controller.go [hunk 1]
```diff
@@ -129,67 +97,131 @@ type ProvisionController struct {
...
+}
 
-	// Map of failed claims
-	failedClaimsStats      map[types.UID]int
-	failedClaimsStatsMutex *sync.Mutex
+// ExponentialBackOffOnError determines whether to exponentially back off from
+// failures of Provision and Delete. Defaults to true.
+func ExponentialBackOffOnError(exponentialBackOffOnError bool) func(*ProvisionController) error {
+	return func(c *ProvisionController) error {
+		c.runningOperations = goroutinemap.NewGoRoutineMap(exponentialBackOffOnError)
```
Comment (xingzhou) [inline lib/controller/controller.go]: @wongma7, do you think it is good that we explicitly add a set of `SetXXX` method in the controller identifying that these attributes can be changed/updated during controller runtime. And for those can not modified ones, if the provisioner called these option methods again in controller runtime, the option method can detect and throw errors.

Comment (wongma7) [inline lib/controller/controller.go]: yes, I'll add SetFailedProvisionThreshold and SetFailedDeleteThreshold. The other attributes however, I don't think can be changed during runtime. And I suppose I could error the options methods: say, after controller Run is called, a flag is set, and if that flag == true all the 'options' functions throw errors.

Comment (wongma7) [inline lib/controller/controller.go]: It's so verbose I am considering not throwing errors if called in controller runtime at all. It should be understood that these are functional options and not something to call during runtime IMO, I will think baout it some more.

**Pull request merged**